@sethrj do you know what I'm missing?
```
/home/runner/work/celeritas/celeritas/src/celeritas/io/EventReader.cc:46:26: error: no member named 'skip' in 'HepMC3::Reader'
            temp_reader->skip(1);
```

--- a/src/accel/HepMC3PrimaryGenerator.cc
+++ b/src/accel/HepMC3PrimaryGenerator.cc
@@ -115,11 +115,24 @@ HepMC3PrimaryGenerator::HepMC3PrimaryGenerator(std::string const& filename)
         SPReader temp_reader = open_hepmc3(filename);
         CELER_ASSERT(temp_reader);
         size_type result = 0;
-        while (!temp_reader->failed())
+#if HEPMC3_VERSION_CODE < 3002000
+        HepMC3::GenEvent evt;
+        temp_reader->read_event(evt);
+#else
+        temp_reader->skip(0);
+#endif
+        CELER_VALIDATE(!temp_reader->failed(),
+                       << "event file '" << filename
+                       << "' did not contain any events");
+        do
         {
-            temp_reader->skip(1);
             result++;
-        }
+#if HEPMC3_VERSION_CODE < 3002000
+            temp_reader->read_event(evt);
+#else
+            temp_reader->skip(1);
+#endif
+        } while (!temp_reader->failed());
         CELER_LOG(debug) << "HepMC3 file has " << result << " events";
         return result;
     }();

--- a/src/celeritas/io/EventReader.cc
+++ b/src/celeritas/io/EventReader.cc
@@ -41,11 +41,24 @@ EventReader::EventReader(std::string const& filename,
         SPReader temp_reader = open_hepmc3(filename);
         CELER_ASSERT(temp_reader);
         size_type result = 0;
-        while (!temp_reader->failed())
+#if HEPMC3_VERSION_CODE < 3002000
+        HepMC3::GenEvent evt;
+        temp_reader->read_event(evt);
+#else
+        temp_reader->skip(0);
+#endif
+        CELER_VALIDATE(!temp_reader->failed(),
+                       << "event file '" << filename
+                       << "' did not contain any events");
+        do
         {
-            temp_reader->skip(1);
             result++;
-        }
+#if HEPMC3_VERSION_CODE < 3002000
+            temp_reader->read_event(evt);
+#else
+            temp_reader->skip(1);
+#endif
+        } while (!temp_reader->failed());
         CELER_LOG(debug) << "HepMC3 file has " << result << " events";
         return result;
     }();

--- a/test/celeritas/io/EventIO.test.cc
+++ b/test/celeritas/io/EventIO.test.cc
@@ -183,6 +183,42 @@ TEST_P(EventIOTest, write_read)
     }
 }
 
+TEST_P(EventIOTest, edge_case)
+{
+    std::string const ext = this->GetParam();
+    std::string filename = this->make_unique_filename(std::string{"."} + ext);
+
+    // Empty event file
+    {
+        std::ofstream out(filename);
+        out << "HepMC::Version 3.02.02\n"
+            << "HepMC::Asciiv3-START_EVENT_LISTING\n"
+            << "HepMC::Asciiv3-END_EVENT_LISTING";
+        out.close();
+        EXPECT_THROW(EventReader(filename, this->particles()), RuntimeError);
+    }
+
+    // Single event
+    {
+        Primary p;
+        p.particle_id = this->particles()->find(pdg::gamma());
+        p.energy = units::MevEnergy{1};
+        p.position = {0, 0, 0};
+        p.direction = {1, 0, 0};
+        p.track_id = TrackId{0};
+        p.event_id = EventId{0};
+        std::vector<Primary> event(4, p);
+        EventWriter write_event(filename, this->particles());
+        write_event(event);
+    }
+    {
+        EventReader read_event(filename, this->particles());
+        EXPECT_EQ(1, read_event.num_events());
+        EXPECT_EQ(4, read_event().size());
+        EXPECT_TRUE(read_event().empty());
+    }
+}
+
 INSTANTIATE_TEST_SUITE_P(EventIO,
                          EventIOTest,
                          testing::Values("hepmc3", "hepmc2", "hepevt"));

--- a/test/celeritas/io/EventIO.test.cc
+++ b/test/celeritas/io/EventIO.test.cc
@@ -5,6 +5,8 @@
 //---------------------------------------------------------------------------//
 //! \file celeritas/io/EventIO.test.cc
 //---------------------------------------------------------------------------//
+#include <fstream>
+
 #include "celeritas_config.h"
 #include "celeritas/io/EventReader.hh"
 #include "celeritas/io/EventWriter.hh"

